### ISSUE TITLE
Pydantic false positive: overrides a field of the same name but is missing a default value

### ISSUE THREAD

**Issue by sk-** [microsoft/pyright#10347]
When overriding a field with a default in a frozen pydantic model we get the error: `overrides a field of the same name but is missing a default value`.

This was already reported in https://github.com/microsoft/pyright/issues/8766, however the example given in https://github.com/microsoft/pyright/issues/8766#issuecomment-2291565896 is unfortunately not correct.
This is because pydantic does not propogate defaults from base classes, opposed to what dataclasses do, which is the justification given in https://github.com/microsoft/pyright/issues/7702 to add such check.

The code below
```python
import pydantic

class Fiz(pydantic.BaseModel, frozen=True):
    fiz: int
    buz: int | None = None

class Buz(Fiz, frozen=True):
    buz: int

Buz(fiz=1)
```

raises the following error at runtime

```
Traceback (most recent call last):
  File "test_dc2.py", line 13, in <module>
    Buz(fiz=1)
  File ".venv/lib/python3.12/site-packages/pydantic/main.py", line 253, in __init__
    validated_self = self.__pydantic_validator__.validate_python(data, self_instance=self)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pydantic_core._pydantic_core.ValidationError: 1 validation error for Buz
buz
  Field required [type=missing, input_value={'fiz': 1}, input_type=dict]
    For further information visit https://errors.pydantic.dev/2.11/v/missing
```

whereas the equivalent code using dataclasses does not raise an error but generates an invalid object from a typing perspective.

```python
import dataclasses


@dataclasses.dataclass(frozen=True)
class Fiz:
    fiz: int
    buz: int | None = None


@dataclasses.dataclass(frozen=True)
class Buz(Fiz):
    buz: int


print(Buz(fiz=1))
```

prints

```
Buz(fiz=1, buz=None) # buz should be an int
```

I checked both [PEP 681: Data Class Transforms](https://peps.python.org/pep-0681/) and [PEP 557: Data Classes](https://peps.python.org/pep-0557/) looking for what the behavior should be for defaults when inheritance is involved but unfortunately I could not find anything.

Additionally, I tested the behavior in the `attrs` package, and they have the same behavior as `pydantic`.

```python
>>> from attrs import define
>>>
>>> @define
... class A:
...   x: int | None = 3
...
>>> @define
... class B(A):
...   x: int
...
>>> A()
A(x=3)
>>> B()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: B.__init__() missing 1 required positional argument: 'x'
```

Given the different behaviors of dataclasses and attrs and pydantic models, would it make sense to add a new option to PEP 681 to signal how defaults are treated in the presence of inheritance and overriding.

This [line](https://github.com/python/cpython/blob/c780698e9b8416633402657ee2846c0179705e9b/Lib/dataclasses.py#L768) in dataclasses is responsible for getting the default from base classes.

**Comment by erictraut** [comment]
You mentioned in the title that pyright is reporting a "false positive" here, but your example demonstrates that pydantic raises an exception at runtime. Wouldn't you want to see an error from your static type checker in this case? For that reason, I don't think this is a false positive.

As you point out, the stdlib dataclass doesn't raise an exception here, but it generates an object that violates the type definition. So once again, I think a static type checker error is justified here.

Arguably, it's possible to create a scenario where the stdlib dataclass would _not_ generate an object that violates the type definition, but I think it's reasonable for a static type checker to treat this as an error even in this situation given that the author probably doesn't intend to rely on this unfortunate behavior of dataclasses.

```python
@dataclass(frozen=True)
class Fiz:
    fiz: int
    buz: int = 0

@dataclass(frozen=True)
class Buz(Fiz):
    buz: int
```

The easy fix in all of these cases is to provide a default value in the subclass.

```python
class Buz(Fiz, frozen=True):
    buz: int = 0
```

> would it make sense to add a new option to PEP 681 to signal how defaults are treated in the presence of inheritance and overriding.

PEPs are point-in-time proposals, so it wouldn't make sense to update the PEP. However, the contents of the PEP have been incorporated into the Python typing spec, which is a living document. You're welcome to propose changes or updates to the Python typing spec. The full process is described [here](https://github.com/python/typing-council#decisions). A good starting point is to start a discussion in the [Python typing forum](https://discuss.python.org/c/typing/32).

**Comment by sk-** [comment]
It is a false positive for pydantic because it is completely valid (without any type checking issues), to override a field without specifying a default.

The error shown at runtime by pydantic means that an argument is missing. So pyright should treat the overriden field as any other field without a default and raise the error

```
error: Argument missing for parameter "buz" (reportCallIssue)
```

The main difference between pydantic and stdlib is that pydantic DOES NOT inherit default values from base classes, which is the issue i'm raising here.

> PEPs are point-in-time proposals

This does not seem to be a set in stone rule as witnessed by the changes to PEP 484. See https://github.com/python/peps/pull/689/files

**Comment by erictraut** [comment]
> This does not seem to be a set in stone rule as witnessed by the changes to PEP 484

Yes, in rare circumstances, PEPs were modified in the past to provide clarifications but not to introduce new functionality. The typing spec is now the last word on type system behaviors, and we have an established process to amend the typing spec.

> So pyright should treat the overriden field as any other field without a default and raise the error

Pyright has no specialized knowledge of pydantic, attrs or any other third-party library — nor should it. Type checking behaviors in this space are dictated by the `dataclass_transform` mechanism and the associated spec. It says that `dataclass_transform` implies standard dataclass behaviors. I can point to many (edge) cases where third-party libraries unfortunately deviate from the standard dataclass behavior. Most of these were not intentional but were side effects of implementation decisions.

I'm going to close this issue because I think pyright is working correctly based on the spec. 

Here are some suggested ways you might proceed:
1. Update your code to specify a default value in inherited classes. This will give you consistent behaviors regardless of whether you're using a stdlib dataclass, pydantic, or attrs.
2. Propose an update to the python typing spec to extend the functionality of `dataclass_transform` so this particular behavior can be specified by pydantic or other libraries.
3. See if the maintainers of pydantic are willing to change the library so it matches the behavior of stdlib dataclasses in this regard. This might be infeasible if many pydantic users are relying on the current behavior.
